### PR TITLE
Restart policy on the containers

### PR DESCRIPTION
**Pull request by vosdev**
Related Issue:

## New Behavior

A restart policy will have the containers starting up automatically unless stopped by hand. After a restart of the docker daemon, the server itself or a process crash, the application will start up again.

## Contrast to Current Behavior

Currently the docker containers will not start up automatically after a docker daemon, server restart or application/process crash. This will require manual intervention (docker-compose up -d) to get the application running again.

## Discussion: Benefits and Drawbacks

With this change, rebooting your server or docker daemon will no longer leave you with an application that is not running.

## Proposed Release Note Entry

NetBox will now automatically start at startup.

## Double Check

* [x] I have read the comments and followed the PR template.
* [x] I have explained my PR according to the information in the comments.
* [x] My PR targets the `develop` branch.

### CODE REVIEW

File: docker-compose.test.yml [hunk 1]
```diff
@@ -2,6 +2,7 @@ version: '3.4'
 services:
   netbox:
     image: ${IMAGE-netboxcommunity/netbox:latest}
+    restart: unless-stopped
```
Comment (cimnine) [inline docker-compose.test.yml]: I'm hesitant to add these to the testing compose file. The reason is that I believe that during testing it should be considered a bug if either service crashes. So I'm keen to learn your your thoughts for adding it to this file.

File: docker-compose.yml [hunk 2]
```diff
@@ -2,6 +2,7 @@ version: '3.4'
 services:
   netbox: &netbox
     image: netboxcommunity/netbox:${VERSION-latest}
+    restart: unless-stopped
```
Comment (cimnine) [inline docker-compose.yml:5]: Why did you go with `unless-stopped` instead of `on-failure`?

**Review by cimnine — COMMENTED** [review]
Thank you for your PR. I guess since some people actually use the docker-compose setup to run this as a productive setup this changes make sense. Please see my two questions.

File: docker-compose.test.yml (hunk 1 above)
Comment (vosdev) [reply docker-compose.test.yml]: The policy on the test can be left out :) My reasoning to add it was to keep it identical to the production version.

File: docker-compose.yml (hunk 2 above)
Comment (vosdev) [reply docker-compose.yml:5]: Behaviour with `unless-stopped` is different compared to `on-failure`. 
With `unless-stopped`:
1. killing a process or having a process exit != 0 will result in docker restarting the container.
2. Restarting docker daemon results in docker starting the containers
3. Restarting your server results in docker starting the containers
4. `docker-compose stop` results in docker not starting your containers after a docker daemon restart or server restart

`on-failure` does 1, and 4 in a similar fashion. But behaviour for 2 and 3 is different. 

Example of `on-failure` behaviour:
```
netbox-docker git/release *
root @ moby # docker-compose ps
            Name                           Command               State    Ports
---------------------------------------------------------------------------------
netbox-docker_netbox-worker_1   python3 /opt/netbox/netbox ...   Up
netbox-docker_netbox_1          /opt/netbox/docker-entrypo ...   Up
netbox-docker_nginx_1           nginx -c /etc/netbox-nginx ...   Up      80/tcp
netbox-docker_postgres_1        docker-entrypoint.sh postgres    Up      5432/tcp
netbox-docker_redis-cache_1     docker-entrypoint.sh sh -c ...   Up      6379/tcp
netbox-docker_redis_1           docker-entrypoint.sh sh -c ...   Up      6379/tcp

netbox-docker git/release *
root @ moby # docker container inspect netbox-docker_netbox_1 --format "{{.HostConfig.RestartPolicy.Name}}"
on-failure

netbox-docker git/release *
root @ moby # systemctl restart docker

netbox-docker git/release *
root @ moby # docker-compose ps
            Name                           Command               State    Ports
-------------------------------------------------------------------------------
netbox-docker_netbox-worker_1   python3 /opt/netbox/netbox ...   Exit 0
netbox-docker_netbox_1          /opt/netbox/docker-entrypo ...   Exit 0
netbox-docker_nginx_1           nginx -c /etc/netbox-nginx ...   Exit 0
netbox-docker_postgres_1        docker-entrypoint.sh postgres    Exit 0
netbox-docker_redis-cache_1     docker-entrypoint.sh sh -c ...   Exit 0
netbox-docker_redis_1           docker-entrypoint.sh sh -c ...   Exit 0
```


Example of `unless-stopped` behaviour:
```
netbox-docker git/release *
root @ moby # docker-compose up -d
Recreating netbox-docker_redis_1       ... done
Recreating netbox-docker_redis-cache_1   ... done
Recreating netbox-docker_postgres_1    ... done
Recreating netbox-docker_netbox-worker_1 ... done
Recreating netbox-docker_netbox_1        ... done
Recreating netbox-docker_nginx_1         ... done

netbox-docker git/release *
root @ moby # docker-compose ps
            Name                           Command               State    Ports
---------------------------------------------------------------------------------
netbox-docker_netbox-worker_1   python3 /opt/netbox/netbox ...   Up
netbox-docker_netbox_1          /opt/netbox/docker-entrypo ...   Up
netbox-docker_nginx_1           nginx -c /etc/netbox-nginx ...   Up      80/tcp
netbox-docker_postgres_1        docker-entrypoint.sh postgres    Up      5432/tcp
netbox-docker_redis-cache_1     docker-entrypoint.sh sh -c ...   Up      6379/tcp
netbox-docker_redis_1           docker-entrypoint.sh sh -c ...   Up      6379/tcp

netbox-docker git/release *
root @ moby # docker container inspect netbox-docker_netbox_1 --format "{{.HostConfig.RestartPolicy.Name}}"
unless-stopped

netbox-docker git/release *
root @ moby # systemctl restart docker

netbox-docker git/release *
root @ moby # docker-compose ps
            Name                           Command               State    Ports
---------------------------------------------------------------------------------
netbox-docker_netbox-worker_1   python3 /opt/netbox/netbox ...   Up
netbox-docker_netbox_1          /opt/netbox/docker-entrypo ...   Up
netbox-docker_nginx_1           nginx -c /etc/netbox-nginx ...   Up      80/tcp
netbox-docker_postgres_1        docker-entrypoint.sh postgres    Up      5432/tcp
netbox-docker_redis-cache_1     docker-entrypoint.sh sh -c ...   Up      6379/tcp
netbox-docker_redis_1           docker-entrypoint.sh sh -c ...   Up      6379/tcp
```


Reason for going `unless-stopped` is that you do not want your application to be dead after a docker daemon restart or server reboot.

File: docker-compose.test.yml (hunk 1 above)
Comment (vosdev) [reply docker-compose.test.yml]: I have removed the policy on the test compose file

**Comment by cimnine** [comment]
What I believe is a problem with this PR is that it makes this project look like it's _plug-and-play_ for production use. But this project's default configurations are not meant to be used as they are _for production_.

Therefore I am hesitant about the change you propose. Because for a productive setup at least the security should be enhanced, as it is described [in the respective wiki page](https://github.com/netbox-community/netbox-docker/wiki/Configuration).

My suggestion is to include the following YAML snippet on [the wiki page _Deployment_](https://github.com/netbox-community/netbox-docker/wiki/Deployment). It's a minimal `docker-compose.override.yml` file to set the restart policies. The page would also explain what the restart policy `unless-stopped` means in practice.

That minimal `docker-compose.override.yml` file would look like this:

```yml
version: '3.4'
services:
  netbox: &netbox
    restart: unless-stopped
  netbox-worker:
    restart: unless-stopped
  nginx:
    restart: unless-stopped
  postgres:
    restart: unless-stopped
  redis:
    restart: unless-stopped
  redis-cache:
    restart: unless-stopped
```

Do you think that would cover your intention?

**Comment by vosdev** [comment]
I'm okay with that.

A custom docker-compose.yml is the preferred method of making local additions to upstream docker-compose.yml's. 
`docker-compose` will merge the dictionaries when fed multiple docker-compose.yml's as such: `docker-compose -f docker-compose.yml -f docker-compose.override.yml up -d`

**Comment by cimnine** [comment]
The good thing about the specific file called `docker-compose.override.yml` is that it's automatically loaded by `docker-compose`. As https://docs.docker.com/compose/extends/#understanding-multiple-compose-files explains:

> By default, Compose reads two files, a `docker-compose.yml` and an optional `docker-compose.override.yml` file. By convention, the `docker-compose.yml` contains your base configuration. The override file, as its name implies, can contain configuration overrides for existing services or entirely new services.

**Comment by cimnine** [comment]
I've documented your suggestion in the wiki. See https://github.com/netbox-community/netbox-docker/wiki/Deployment#docker-compose-based-deployments

**Comment by vosdev** [comment]
> The good thing about the specific file called `docker-compose.override.yml` is that it's automatically loaded by `docker-compose`. As https://docs.docker.com/compose/extends/#understanding-multiple-compose-files explains:
> 
> > By default, Compose reads two files, a `docker-compose.yml` and an optional `docker-compose.override.yml` file. By convention, the `docker-compose.yml` contains your base configuration. The override file, as its name implies, can contain configuration overrides for existing services or entirely new services.

Today I learned! Thank you :)

> I've documented your suggestion in the wiki. See https://github.com/netbox-community/netbox-docker/wiki/Deployment#docker-compose-based-deployments

Looks good!

**Pull request closed without merging**